crates/accelerate/src/target_transpiler/mod.rs

```diff
@@ -966,6 +897,179 @@ impl Target {
+                for qarg in props_map.keys() {
+                    if let QargsRef::Concrete(qarg_slice) = qarg.as_ref() {
+                        if qarg_slice.len() != instruction.num_qubits() as usize {
+                            return Err(TargetError::QargsMismatch {
+                                instruction: name,
+                                arguments: qarg.clone(),
+                            });
+                        }
+                        self.num_qubits = Some(self.num_qubits.unwrap_or_default().max(
+                            qarg_slice.iter().fold(0, |acc, x| {
+                                if acc > x.index() {
+                                    acc
+                                } else {
+                                    x.index()
+                                }
+                            }) + 1,
+                        ));
+                    }
+                    if let Some(Some(value)) = self.qarg_gate_map.get_mut(&qarg.as_ref()) {
+                        value.insert(name.to_string());
+                    } else {
+                        self.qarg_gate_map
+                            .insert(qarg.clone(), Some(HashSet::from_iter([name.to_string()])));
+                    }
+                }
+            }
+        }
+        self._gate_name_map.insert(name.to_string(), instruction);
+        self.gate_map.insert(name.to_string(), props_map);
+        self.non_global_basis = None;
+        self.non_global_strict_basis = None;
+        Ok(())
+    }
+
+    /// Update the property object for an instruction qarg pair already in the [Target].
+    ///
+    /// # Arguments
+    ///
+    /// * `instruction` - The instruction's name within this instance.
+    /// * `qargs` - A collection of [PhysicalQubit] or an instance of [Qargs::Global]
+    ///   that the instruction operated on.
+    ///
+    /// # Returns
+    ///
+    /// * `Ok`: if the instruction property is successfully updated.
+    /// * `Err`: (if neither the instruction name or qarg aren't found) [TargetError].
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use qiskit_accelerate::target_transpiler::{Target, InstructionProperties, Qargs};
+    /// use qiskit_circuit::operations::StandardGate;
+    /// use qiskit_accelerate::nlayout::PhysicalQubit;
+    /// use indexmap::IndexMap;
+    ///
+    /// let mut target = Target::default();
+    /// target.add_instruction(
+    ///     StandardGate::X.into(),
+    ///     &[],
+    ///     None,
+    ///     Some(IndexMap::from_iter([([PhysicalQubit(0)].into(), None)])),
+    /// );
+    /// let result = target.update_instruction_properties("x", &[PhysicalQubit(0)], Some(InstructionProperties::new(Some(0.0001), Some(0.0002))));
+    ///
+    /// assert!(matches!(result, Ok(())));
+    /// ```
+    pub fn update_instruction_properties<'a, T>(
+        &mut self,
+        instruction: &'a str,
+        qargs: T,
+        properties: Option<InstructionProperties>,
```

Comment:
Do we need the `Option` here? Like I realize the data model supports `None` to indicate an ideal instruction with no properties, but I wonder if in practice this is something we should be supporting on an update method. The python space typing indicated `None` wasn't an allowed type only `InstructionProperties`.

Reply:
That's a good question. Getting rid of the option seems logical. However, what if the user were to update the properties of a global/variadic instruction temporarily, and then back to its original value `None`? What could the user do in this case and is this supported behavior?

Reply:
I'm fine leaving it. I don't actually think that's a thing people will do in practice. The handling of globals is ill defined especially with properties set and it's not commonly used at all. That was not part of the original design intent of the target, and the fact it was possible was an implementation mistake that leaves us with undefined behavior. This is why I asked you to open #14268 so we can fix this hole in the API for 3.0 (regardless of what we do it will probably be backwards incompatible).